### ISSUE TITLE
Using different listeners and advertised addresses produces a fatal error

### ISSUE THREAD

**Issue by G-Goldstein** [grepplabs/kafka-proxy#93]
I'm trying to run kafka-proxy in a Kubernetes cluster, proxying on to ConfluentCloud. In order to do this, I'll need to be able to tell Kafka clients about the listener address for the kafka-proxy, which is a Kubernetes Service. At the same time, I need kafka-proxy to bind to local ports in order to function.

I read in the body of another issue that the `--external-server-mapping` can take three subparts, separated by commas: The first is the address to proxy on to, the second is the listener address, and the third is an optional advertised address. That sounds like exactly what I need, so I've set that up in the args:

      - --external-server-mapping
      - b0-pkc-l6wr6.europe-west2.gcp.confluent.cloud:9092,0.0.0.0:9100,kafka-proxy:9100

But it seems this isn't acceptable, because on startup I'm immediately met with this fatal error, which kills the process:

    time="2021-11-01T14:47:19Z" level=info msg="Starting kafka-proxy version v0.3.0"
    time="2021-11-01T14:47:19Z" level=info msg="Bootstrap server pkc-l6wr6.europe-west2.gcp.confluent.cloud:9092 advertised as 0.0.0.0:9092"
    time="2021-11-01T14:47:19Z" level=fatal msg="external server mapping has different listener and advertised addresses {b0-pkc-l6wr6.europe-west2.gcp.confluent.cloud:9092 0.0.0.0:9100 kafka-proxy:9100}"

I'm sure there's something I'm not understanding here because this error message is obviously deliberate, but it seems as though this optional third parameter only works if it shares its value with the second parameter, making it useless. I think I need to use 0.0.0.0 for the kafka-proxy to bind to the address, but I can't pass 0.0.0.0 on to the kafka clients because it isn't valid within their own containers. How should I set this up?

**Comment by everesio** [comment]
You should provide only 2 args in the `external-server-mapping` and skipping the listener bind part e.g. `b0-pkc-l6wr6.europe-west2.gcp.confluent.cloud:9092,kafka-proxy:9100`. The difference between `external-server-mapping` and `bootstrap-server-mapping` is that for external server configuration a TCP listener is not started and it is only used for address mapping.

Following snippet gives you an example how you can generate STS for kubernetes (the same can be done for service) using helm

```
{{- $currentScope := .Values }}
{{ range $i := until (.Values.brokers | int) }}

---
apiVersion: apps/v1
kind: StatefulSet
metadata:
   name: kafka-gateway-{{ $i }}
   namespace: kafka-gateway

...

args:
  - 'server'
  - '--log-format=json'
  - '--bootstrap-server-mapping=kafka-{{ $i }}:9093,0.0.0.0:32400,kafka-gateway-{{ $i }}.{{ $currentScope.dnszone }}:32400'
  {{- range $j := until (int $currentScope.brokers) }}
  - '--external-server-mapping=kafka-{{ $j }}:9093,kafka-gateway-{{ $j }}.{{ $currentScope.dnszone }}:32400'
  {{- end }}
{{end}}

```

**Comment by G-Goldstein** [comment]
Thanks, this setup is what I needed and my clients are able to talk through the proxy now! I must admit that I've copied your example but don't fully understand `bootstrap-server-mapping` and `external-server-mapping`. I expected to only use `bootstrap-server-mapping` for the bootstrap servers and not also the brokers. I'm also unsure as to why I need both entries, or what would happen if I removed the `external-server-mapping` now.

I'm not looking to get these answers here and this question can be closed. I'm just aiming to provide some feedback for the documentation.
